grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/GreedyTraversalPlan.java

```diff
@@ -136,6 +175,122 @@ public static GraqlTraversal createTraversal(PatternAdmin pattern, EmbeddedGrakn
+        labelVarTypeMap.values().stream().distinct().forEach(
+                type -> getAllPossibleRelationships(relationshipMap, type));
+
+        relationshipRolePlayerMap.asMap().forEach((relationshipVar, rolePlayerVars) -> {
+            Iterator<Var> rolePlayerVarIterator = rolePlayerVars.iterator();
+
+            Var firstRolePlayer = rolePlayerVarIterator.next();
+            while (!instanceVarTypeMap.containsKey(firstRolePlayer)) {
+                firstRolePlayer = rolePlayerVarIterator.next();
+            }
+
+            Set<Type> possibleRelationShipTypes = getAllPossibleRelationShipTypesOfARolePlayer(
+                    firstRolePlayer, instanceVarTypeMap, relationshipMap);
+            while (rolePlayerVarIterator.hasNext()) {
+                Var nextRolePlayer = rolePlayerVarIterator.next();
+                if (instanceVarTypeMap.containsKey(nextRolePlayer)) {
+
+                    // compute the intersection of possible relationship types
+                    possibleRelationShipTypes.retainAll(getAllPossibleRelationShipTypesOfARolePlayer(
+                            nextRolePlayer, instanceVarTypeMap, relationshipMap));
+                }
+            }
+
+            //TODO: if possibleRelationShipTypes here is empty, the query will not match any data
+            if (possibleRelationShipTypes.size() == 1) {
+
+                Type relationshipType = possibleRelationShipTypes.iterator().next();
+                Label label = relationshipType.getLabel();
+                Var labelVar = var();
+                allFragments.add(Fragments.label(LabelProperty.of(label), labelVar, ImmutableSet.of(label)));
+                allFragments.addAll(EquivalentFragmentSets.isa(IsaProperty.of(label(label).admin()),
+                        relationshipVar, labelVar, relationshipType.isImplicit()).fragments());
+            }
+        });
+    }
+
+    private static Set<Type> getAllPossibleRelationShipTypesOfARolePlayer(
+            Var rolePlayer, Multimap<Var, Type> instanceVarTypeMap, Map<Type, Set<RelationshipType>> relationshipMap) {
+
+        Iterator<Type> typeIterator = instanceVarTypeMap.get(rolePlayer).iterator();
+        Type firstRolePlayerType = typeIterator.next(); // the role player here has at least one type
+
+        Set<Type> relationshipTypes = new HashSet<>(relationshipMap.get(firstRolePlayerType));
+
+        // we can stop doing intersection when relationshipTypes is empty
+        while (typeIterator.hasNext() && !relationshipTypes.isEmpty()) {
+            Type nextRolePlayerType = typeIterator.next();
+            relationshipTypes.retainAll(relationshipMap.get(nextRolePlayerType));
+        }
+
+        //TODO: if relationshipTypes here is empty, the query will not match any data
+        return relationshipTypes;
+    }
+
+    private static void getAllPossibleRelationships(Map<Type, Set<RelationshipType>> relationshipMap, Type metaType) {
```

Comment:
Since this mutates the collection you give it, maybe call it `addAllPossibleRelationships`.

Also, is the second argument always a meta-type?